Update mapping-cognito-identity-id-and-user-pool-id.md

meech-ward opened the pull request:
Full disclosure, I gotta run to work in a few minutes and don't have time to check this out in-depth, but:

I couldn't find any documentation on AWS's end about this, it just seems that `event.requestContext.identity` is always undefined when using `iam` as an authorizer. My guess is that they switched to `event.requestContext.authorizer.iam` at some point. 

There's probably a better way of doing this, but here's my solution this morning: `const authProvider = event.requestContext.authorizer.iam.cognitoIdentity.amr.findLast(ref => ref.includes(':'))`


For reference here's my output of `event.requestContext.authorizer.iam`

```
event.requestContext.authorizer.iam {
  accessKey: 'accessKey',
  accountId: 'accountId',
  callerId: 'callerId:CognitoIdentityCredentials',
  cognitoIdentity: {
    amr: [
      'authenticated',
      'cognito-idp.us-west-2.amazonaws.com/us-west-2_abcdef',
      'cognito-idp.us-west-2.amazonaws.com/us-west-2_abcdef:CognitoSignIn:user_pool_user_id'
    ],
    identityId: 'us-west-2:identityId',
    identityPoolId: 'us-west-2:identityPoolId'
  },
  principalOrgId: null,
  userArn: 'arn:aws:sts::me:assumed-role/sam-cognito-test-API-AuthIdentityPoolAuthRole_role/CognitoIdentityCredentials',
  userId: 'userId:CognitoIdentityCredentials'
}
```

jayair commented:
Oh I think this changed with the API Gateway v1 to HTTP API move? Can somebody double check?

meech-ward commented:
@jayair that's right

## HTTP API Gateway

```js
const api = new Api(stack, "Api", {
  defaults: {
    authorizer: "iam",
  },
  // ...
});
```

The output of `event.requestContext.authorizer.iam`:

```js
event.requestContext.authorizer.iam: {
  accessKey: 'accessKey',
  accountId: 'accountId',
  callerId: 'callerId:CognitoIdentityCredentials',
  cognitoIdentity: {
    amr: [
      'authenticated',
      'cognito-idp.us-west-2.amazonaws.com/us-west-2_abcdef',
      'cognito-idp.us-west-2.amazonaws.com/us-west-2_abcdef:CognitoSignIn:user_pool_user_id'
    ],
    identityId: 'us-west-2:identityId',
    identityPoolId: 'us-west-2:identityPoolId'
  },
  principalOrgId: null,
  userArn: 'arn:aws:sts::me:assumed-role/sam-cognito-test-API-AuthIdentityPoolAuthRole_role/CognitoIdentityCredentials',
  userId: 'userId:CognitoIdentityCredentials'
}
```

To get the user pool user id: 

```js
export async function main(event, context) {
  const authProvider = event.requestContext.authorizer.iam.cognitoIdentity.amr.findLast(ref => ref.includes(':'))
  const parts = authProvider.split(':');
  const userPoolIdParts = parts[parts.length - 3].split('/');
  const userPoolId = userPoolIdParts[userPoolIdParts.length - 1];
  const userPoolUserId = parts[parts.length - 1];
  // ...
}
```



## REST API Gateway

```js
const api = new ApiGatewayV1Api(stack, "ApiRest", {
  defaults: {
    authorizer: "iam",
  },
  // ...
});
```


The output of `event.requestContext.identity`

```js
event.requestContext.identity: {
  cognitoIdentityPoolId: 'us-west-2:identityPoolId',
  accountId: 'accountId',
  cognitoIdentityId: 'us-west-2:identityId',
  caller: 'caller:CognitoIdentityCredentials',
  sourceIp: 'sourceIp',
  principalOrgId: null,
  accessKey: 'accessKey',
  cognitoAuthenticationType: 'authenticated',
  cognitoAuthenticationProvider: 'cognito-idp.us-west-2.amazonaws.com/user_pool_id,cognito-idp.us-west-2.amazonaws.com/user_pool_id:CognitoSignIn:user_pool_user_id',
  userArn: 'arn:aws:sts::user:assumed-role/sam-sst-app-API-AuthIdentityPoolAuthRole/CognitoIdentityCredentials',
  user: 'caller:CognitoIdentityCredentials'
}
```

To get the user pool user id: 

```js
export async function main(event, context, callback) {
  const authProvider = event.requestContext.identity.cognitoAuthenticationProvider;
  const parts = authProvider.split(':');
  const userPoolIdParts = parts[parts.length - 3].split('/');
  const userPoolId = userPoolIdParts[userPoolIdParts.length - 1];
  const userPoolUserId = parts[parts.length - 1];
  // ...
}
```

jayair commented:
Ah. Nice find.
